tmail-backend/jmap/extensions-cassandra/src/main/scala/com/linagora/tmail/james/jmap/firebase/CassandraFirebaseSubscriptionDAO.scala

```diff
@@ -0,0 +1,106 @@
+import com.datastax.oss.driver.api.querybuilder.QueryBuilder.{bindMarker, deleteFrom, insertInto, selectFrom}
+import com.linagora.tmail.james.jmap.model.{DeviceClientId, FirebaseDeviceToken, FirebaseSubscription, FirebaseSubscriptionExpiredTime, FirebaseSubscriptionId}
+import javax.inject.Inject
+import org.apache.james.backends.cassandra.components.CassandraModule
+import org.apache.james.backends.cassandra.utils.CassandraAsyncExecutor
+import org.apache.james.core.Username
+import org.apache.james.jmap.api.change.TypeStateFactory
+import org.apache.james.jmap.api.model.TypeName
+import reactor.core.scala.publisher.{SFlux, SMono}
+
+import scala.collection.immutable.Seq
+import scala.jdk.javaapi.CollectionConverters
+
+object CassandraFirebaseSubscriptionTable {
+  val TABLE_NAME = "firebase_subscription"
+  val USER = "user"
+  val DEVICE_CLIENT_ID = "device_client_id"
+  val ID = "id"
+  val EXPIRES = "expires"
+  val TYPES = "types"
+  val TOKEN = "device_token"
+
+  val MODULE: CassandraModule = CassandraModule.table(TABLE_NAME)
+    .comment("Hold user firebase push subscriptions data")
+    .statement(statement => types => statement
+      .withPartitionKey(USER, TEXT)
+      .withClusteringColumn(DEVICE_CLIENT_ID, TEXT)
+      .withColumn(TOKEN, TEXT)
+      .withColumn(ID, UUID)
+      .withColumn(EXPIRES, TIMESTAMP)
+      .withColumn(TYPES, frozenSetOf(TEXT)))
+    .build
+}
+
+class CassandraFirebaseSubscriptionDAO @Inject()(session: CqlSession, typeStateFactory: TypeStateFactory) {
+
+  import CassandraFirebaseSubscriptionTable._
+
+  private val executor: CassandraAsyncExecutor = new CassandraAsyncExecutor(session)
+
+  private val insert: PreparedStatement = session.prepare(insertInto(TABLE_NAME)
+    .value(USER, bindMarker(USER))
+    .value(DEVICE_CLIENT_ID, bindMarker(DEVICE_CLIENT_ID))
+    .value(TOKEN, bindMarker(TOKEN))
+    .value(ID, bindMarker(ID))
+    .value(EXPIRES, bindMarker(EXPIRES))
+    .value(TYPES, bindMarker(TYPES))
+    .build())
+
+  private val selectAll: PreparedStatement = session.prepare(selectFrom(TABLE_NAME)
+    .all()
+    .whereColumn(USER).isEqualTo(bindMarker(USER)).build())
+
+  private val deleteOne: PreparedStatement = session.prepare(deleteFrom(TABLE_NAME)
+    .whereColumn(USER).isEqualTo(bindMarker(USER))
+    .whereColumn(DEVICE_CLIENT_ID).isEqualTo(bindMarker(DEVICE_CLIENT_ID))
+    .build())
+
+  def insert(username: Username, subscription: FirebaseSubscription): SMono[FirebaseSubscription] = {
+    val typeNames = CollectionConverters.asJava(subscription.types.map(_.asString()).toSet)
+    val utcInstant = subscription.expires.value.withZoneSameInstant(ZoneOffset.UTC).toInstant
+
+    val insertSubscription = insert.bind()
+      .setString(USER, username.asString)
+      .setString(DEVICE_CLIENT_ID, subscription.deviceClientId.value)
+      .setUuid(ID, subscription.id.value)
+      .setInstant(EXPIRES, utcInstant)
+      .setSet(TYPES, typeNames, classOf[String])
```

Comment:
Please have a static variable for TypeCodec (getting collection type codecs is expensive)

```
    private static final TypeToken<Set<String>> SET_OF_STRINGS = TypeTokens.setOf(String.class);
    private static final TypeCodec<Set<String>> FROZEN_OF_STRINGS_CODEC = CodecRegistry.DEFAULT_INSTANCE.codecFor(DataType.frozenSet(DataType.text()), SET_OF_STRINGS);

```